pydantic/_internal/_validate_call.py

```diff
@@ -97,3 +105,122 @@ def __call__(self, *args: Any, **kwargs: Any) -> Any:
         if self.__return_pydantic_validator__:
             return self.__return_pydantic_validator__(res)
         return res
+
+
+class ValidateCallInfo(TypedDict):
+    validate_return: bool
+    config: ConfigDict | None
+    function: Callable[..., Any]
+    local_namespace: dict[str, Any] | None
+
+
+def validate_call_with_namespace(
+    function: AnyCallableT | None = None,
+    config: ConfigDict | None = None,
+    validate_return: bool = False,
+    local_namespace: dict[str, Any] | None = None,
+) -> AnyCallableT | Callable[[AnyCallableT], AnyCallableT]:
+    def validate(function: AnyCallableT) -> AnyCallableT:
+        if isinstance(function, (classmethod, staticmethod)):
+            name = type(function).__name__
+            raise TypeError(f'The `@{name}` decorator should be applied after `@validate_call` (put `@{name}` on top)')
+
+        validate_call_wrapper = ValidateCallWrapper(function, config, validate_return, local_namespace)
+
+        if inspect.iscoroutinefunction(function):
+
+            @wraps(function)
+            async def wrapper_function(*args, **kwargs):  # type: ignore
+                return await validate_call_wrapper(*args, **kwargs)
+        else:
+
+            @wraps(function)
+            def wrapper_function(*args, **kwargs):
+                return validate_call_wrapper(*args, **kwargs)
+
+        wrapper_function.raw_function = function  # type: ignore
+
+        info = ValidateCallInfo(
+            validate_return=validate_return,
+            config=config,
+            function=function,
+            local_namespace=local_namespace,
+        )
+        wrapper_function.__pydantic_validate_call_info__ = info  # type: ignore
+
+        return wrapper_function  # type: ignore
+
+    if function:
+        return validate(function)
+    else:
+        return validate
+
+
+def _is_wrapped_by_validate_call(obj: object) -> bool:
+    return hasattr(obj, '__pydantic_validate_call_info__')
+
+
+def collect_validate_call_info(namespace: Mapping[str, Any]) -> dict[str, ValidateCallInfo]:
+    return {
+        name: func.__pydantic_validate_call_info__
+        for name, func in namespace.items()
+        if _is_wrapped_by_validate_call(func)
+    }
+
+
+def _update_qualname(function: Callable[..., Any], model: type[BaseModel]) -> None:
+    origin = model.__pydantic_generic_metadata__['origin']
+    if not origin:
+        return
+
+    name = function.__name__
+    qualname = function.__qualname__
+
+    original_postfix = f'{origin.__name__}.{name}'
+    assert qualname.endswith(original_postfix)
+    function.__qualname__ = qualname.replace(original_postfix, f'{model.__name__}.{name}')
+
+
+def update_generic_validate_calls(model: type[BaseModel]) -> None:
```

Comment:
You will have to add a `try..finally` block in case any exception happens, as otherwise the original function will have its attributes mutated.

Also, I'll have to think about it more, but are we certain that every callable that can be used with `validate_call` can have the `__qualname__`/`__annotations__` etc arguments mutated?

Reply:
Ah, I think I forgot to consider callables with various type in many places. Later I should add more tests with `partial`, `wraps` etc. 

Also, #10461 should probably be dealt with first to better recognize what types of functions we may have here.